src/lsp4clj/server.clj

```diff
@@ -112,22 +112,81 @@
-  (let [message-type (coercer/input-message-type message)]
-    (try
+(defn thread-loop [buf-or-n f]
+  (let [ch (async/chan buf-or-n)]
+    (async/thread
       (discarding-stdout
-        (case message-type
-          (:parse-error :invalid-request)
-          (protocols.endpoint/log server :error (format-error-code "Error reading message" message-type))
-          :request
-          (protocols.endpoint/receive-request server context message)
-          (:response.result :response.error)
-          (protocols.endpoint/receive-response server message)
-          :notification
-          (protocols.endpoint/receive-notification server context message)))
-      (catch Throwable e ;; exceptions thrown by receive-response or receive-notification (receive-request catches its own exceptions)
-        (log-error-receiving server e message)))))
+        (loop []
+          (when-let [arg (async/<!! ch)]
+            (f arg)
+            (recur)))))
+    ch))
+
+(defn ^:private dispatch-input
+  "Dispatches messages received on the input-ch based on message type. Returns a
+  channel which will close after the input-ch is closed."
+  [server context input-ch]
+  (let [;; In order to process some requests and (all) notifications in series,
+        ;; the language server sometimes needs to block client-initiated input.
+        ;; If the language server sends requests during that time, it needs to
+        ;; receive responses, even though it's blocking other input. Otherwise,
+        ;; it will end up in a deadlock, where it's waiting to receive a
+        ;; response off the input-ch and the input-ch isn't being read from
+        ;; because the server is blocking input. See
+        ;; https://github.com/clojure-lsp/clojure-lsp/issues/1500.
+
+        ;; The messages all arrive in order on the input-ch so to get to the
+        ;; client's response, we have to queue whatever other messages it's
+        ;; sent. We do that by storing them in a sliding buffer. Because of the
+        ;; sliding buffer:
+        ;; * if the client sends a message which causes the language server to
+        ;;   block, and
+        ;; * if the language server sends a request during that time, and
+        ;; * if the client sends more than 100 other messages between when the
+        ;;   language server started blocking and when the client responds to
+        ;;   the language server's request,
+        ;; * then the client's earliest messages will be dropped.
+        ;; The same is true in reverse.
+
+        ;; We process the client- and language-server-initiated messages in
+        ;; separate threads.
+        ;; * Threads, so the language server can use >!! and so that we can use
+        ;;   (>!! output-ch) to respect back pressure from clients that are slow
+        ;;   to read.
+        ;; * Separate, so one can continue while the other is blocked.
+
+        ;; (Jacob Maine): 100 is picked out of thin air. I have no idea how to
```

Comment:
I think we are safe with 100 messages, I can't see a case where makes sense a buffer bigger than that